src/stdlib_32_bit_fnv_hashes.fypp

```diff
@@ -0,0 +1,126 @@
+!!------------------------------------------------------------------------------
+!! `FNV_1_HASH` and  `FNV_1A_Hash` are translations to Fortran 2008 of the
+!! `FNV-1` and `FNV-1a` hash functions of Glenn Fowler, Landon Curt Noll,
+!! and Phong Vo, that has been released into the public domain. Permission
+!! has been granted, by Landon Curt Noll, for the use of these algorithms
+!! in the Fortran Standard Library. A description of these functions is
+!! available at https://en.wikipedia.org/wiki/Fowler–Noll–Vo_hash_function.
+!!------------------------------------------------------------------------------
+
+!#! Integer kinds to be considered during templating
+#:set INT_KINDS = ["int16", "int32", "int64"]
+
+submodule(stdlib_32_bit_hash_codes) stdlib_32_bit_fnv_hashes
+!! An implementation of the FNV hashes 1 and 1a of Glenn Fowler, Landon Curt
+!! Noll, and Kiem-Phong-Vo,
+!! https://en.wikipedia.org/wiki/Fowler–Noll–Vo_hash_function
+    implicit none
+
+    integer(int_hash), parameter ::                  &
+        offset_basis = int( z'811C9DC5', int_hash ), &
+        prime        = int( z'01000193', int_hash )
+
+contains
+
+    pure module function int8_fnv_1( key ) result(hash_code)
+!! The original FNV-1 8-bit key algorithm.
+        integer(int8), intent(in)     :: key(:)
+        integer(int_hash)             :: hash_code
+
+        integer(int64) :: i
+
+        hash_code = offset_basis
+        do i=1_int64, size(key, kind=int64)
+            hash_code = hash_code * prime
+            if ( little_endian ) then
+                hash_code = ieor( hash_code, &
+                                  transfer( [key(i), 0_int8, 0_int8, 0_int8], &
+                                            0_int_hash ) )
+            else
+                hash_code = ieor( hash_code, &
+                                  transfer( [0_int8, 0_int8, 0_int8, key(i)], &
+                                            0_int_hash ) )
+            end if
+        end do
+
+    end function int8_fnv_1
+
+
+#:for k1 in INT_KINDS
+    pure module function ${k1}$_fnv_1( key ) result(hash_code)
+! A ${k1}$ array key wrapper for the FNV-1 algorithm.
+        integer(${k1}$), intent(in) :: key(:)
+        integer(int_hash)           :: hash_code
+
+        hash_code = int8_fnv_1( transfer( key, 0_int8,                      &
+                                          bytes_${k1}$*                     &
+                                          size( key, kind=int64 ) ) )
+
+    end function ${k1}$_fnv_1
+
+#:endfor
+
+
+    pure module function character_fnv_1( key ) result(hash_code)
```

Comment:
Can we make this elemental (so one can also pass an array of keys)?

Reply:
I added a bunch of comments like this below and then stopped -- but the general question is, for the functions that take scalar arguments, can we make them elemental if possible?

Reply:
In response to your suggestions I have examined the code and documentation and made the following changes.

1. In examining `stdlib_64_bit_hash_codes.fypp` I discovered that I had forgotten to remove references to murmur hash so I removed them.
2. For `fibonacci_hash` and `universaLmult_hash` all definitions/interfaces were made elemental.
3. For `fnv_1`, `fnv_1a`, `nmhash`, `nmhashx`, `water_hash`, and `pengy_hash` functions with character string keys were made elemental.
4. For `spooky_hash` the analysis was more complicated than I had time for. I suspect the hash on integer arrays is pure and on character strings is elemental, but I will need more time to be certain. The documentation, but not the definition/interface had the function as pure so that attribute was removed form the documentation.